devito/pointdata.py

```diff
@@ -0,0 +1,197 @@
+        kwargs['shape'] = (nt, npoint)
+
+        return DenseData.__new__(cls, *args, **kwargs)
+
+    @classmethod
+    def _indices(cls, **kwargs):
+        """Return the default dimension indices for a given data shape
+
+        :param shape: Shape of the spatial data
+        :return: indices used for axis.
+        """
+        dimensions = kwargs.get('dimensions', None)
+        return dimensions or [t, p]
+
+    @property
+    def coefficients(self):
+        """Symbolic expression for the coefficients for sparse point
+        interpolation according to:
+        https://en.wikipedia.org/wiki/Bilinear_interpolation.
+
+        :returns: List of coefficients, eg. [b_11, b_12, b_21, b_22]
+        """
+        # Grid indices corresponding to the corners of the cell
+        x1, y1, z1, x2, y2, z2 = symbols('x1, y1, z1, x2, y2, z2')
+        # Coordinate values of the sparse point
+        px, py, pz = self.point_symbols
+        if self.ndim == 2:
+            A = Matrix([[1, x1, y1, x1*y1],
+                        [1, x1, y2, x1*y2],
+                        [1, x2, y1, x2*y1],
+                        [1, x2, y2, x2*y2]])
+
+            p = Matrix([[1],
+                        [px],
+                        [py],
+                        [px*py]])
+
+        elif self.ndim == 3:
+            A = Matrix([[1, x1, y1, z1, x1*y1, x1*z1, y1*z1, x1*y1*z1],
+                        [1, x1, y2, z1, x1*y2, x1*z1, y2*z1, x1*y2*z1],
+                        [1, x2, y1, z1, x2*y1, x2*z1, y2*z1, x2*y1*z1],
+                        [1, x1, y1, z2, x1*y1, x1*z2, y1*z2, x1*y1*z2],
+                        [1, x2, y2, z1, x2*y2, x2*z1, y2*z1, x2*y2*z1],
+                        [1, x1, y2, z2, x1*y2, x1*z2, y2*z2, x1*y2*z2],
+                        [1, x2, y1, z2, x2*y1, x2*z2, y1*z2, x2*y1*z2],
+                        [1, x2, y2, z2, x2*y2, x2*z2, y2*z2, x2*y2*z2]])
+
+            p = Matrix([[1],
+                        [px],
+                        [py],
+                        [pz],
+                        [px*py],
+                        [px*pz],
+                        [py*pz],
+                        [px*py*pz]])
+        else:
+            error("Interpolation coefficients not implemented for "
+                  "%d dimensions." % self.ndim)
+
+        # Map to reference cell
+        reference_cell = {x1: 0, y1: 0, z1: 0, x2: h, y2: h, z2: h}
+        A = A.subs(reference_cell)
+        return A.inv().T.dot(p)
+
+    @property
+    def point_symbols(self):
+        """Symbol for coordinate value in each dimension of the point"""
+        return symbols('px, py, pz')
+
+    @property
+    def point_increments(self):
+        """Index increments in each dimension for each point symbol"""
+        if self.ndim == 2:
+            return ((0, 0), (0, 1), (1, 0), (1, 1))
+        elif self.ndim == 3:
+            return ((0, 0, 0), (0, 1, 0), (1, 0, 0), (0, 0, 1),
+                    (1, 1, 0), (0, 1, 1), (1, 0, 1), (1, 1, 1))
```

Comment:
AFAIK everything else in core devito is independent of the number of dimensions and with this you are introducing a restriction on the supported dimensionality. Maintaining generality is the scope of every PR.